lib/ModalQueue.tsx

```diff
@@ -0,0 +1,214 @@
+   * 
+   * You can also specify a custom `delay` to a specific modal in the options of `enqueueModal`.
+   * 
+   * Default value: `0`.
+   */
+  minimumDelay = 0;
+
+  /**
+   * Whether this queue should be preserved by `ModalQueuesManager` after all modals in this queue are dismissed.
+   * 
+   * Default value: `false`.
+   */
+  preserveWhenEmpty = false;
+
+  /**
+   * If `true`, this queue puases presenting modals. 
+   * 
+   * Default value: `false`.
+   */
+  paused = false;
+
+  /**
+   * Delay in millisecond before the queue actually starts presenting modals.
+   * 
+   * This delay allows some time to keep multiple linear `enqueueModal` calls
+   * in expected order, based on their priorities.
+   * 
+   * Default value: `10`.
+   */
+  delayBeforeFiring = 10;
+
+  /**
+   * Whether this queue has any modal being presented or to be presented.
+   */
+  get isEmpty() {
+    return this.modals.length < 1;
+  }
+
+  /**
+   * Use `addQueue` method of `ModalQueuesManager` to create and add queues.
+   */
+  constructor(name: string, delegate: ModalQueueDelegate) {
+    this.name = name;
+    this.delegate = delegate;
+  }
+
+  /**
+   * Start the queue to present modals if there are modals to be presented and the queue is not paused.
+   */
+  fire = async () => {
+    if (this.running || this.isEmpty || this.paused) {
+      return;
+    }
+    this.running = true;
+    while(!this.paused && !this.isEmpty) {
+      await this.pop();
+    }
+    this.running = false;
+    this.delegate.onStop();
+  };
+
+  /**
+   * Add a modal to be presented to this queue.
+   * 
+   * @param element the modal to be presented.
+   */
+  push = (element: ModalQueueElement) => {
+    const highs: typeof this.modals = [];
+    const defaults: typeof this.modals = [];
+    const lows: typeof this.modals = [];
+
+    this.modals.forEach(_element => {
+      if (_element.priority === 'high') {
+        highs.push(_element);
+      } else if (_element.priority === 'low') {
+        lows.push(_element);
+      } else {
+        defaults.push(_element);
+      }
+    });
+    if (element.priority === 'high') {
+      highs.push(element);
+    } else if (element.priority === 'low') {
+      lows.push(element);
+    } else {
+      defaults.push(element);
+    }
+    this.modals = highs.concat(defaults).concat(lows);
+
+    setTimeout(() => {
+      this.fire();
+    }, this.delayBeforeFiring);
+  };
+
+  /**
+   * Remove the modal of highest priority from the queue and present it on screen.
+   * 
+   * This method waits for the modal to be dismissed before it returns.
+   */
+  pop = async () => {
+    if (this.isEmpty) {
+      return;
+    }
+    const modal = this.modals[0];
+    const finalDelay = Math.max(modal.delay ?? 0, this.minimumDelay);
+    if (finalDelay > 0) {
+      await (new Promise<void>(resolve => setTimeout(resolve, finalDelay)));
+    }
+    await (new Promise<void>(resolve => modal.present(resolve)));
```

Comment:
I think this can also be simplified as just
```
await (new Promise<void>(resolve => setTimeout(resolve => modal.present(resolve), finalDelay)));
```

Reply:
makes sense, let's leave it as it is then